src/aspire/basis/basis_utils.py

```diff
@@ -320,3 +324,93 @@ def lgwt(ndeg, a, b):
     x = x[ids]
 
     return x, w
+
+
+def j_polynomial(m, n, alpha, beta, x):
+    """
+    The Jacobi polynomials defined in the paper, eq (2), page 6
+
+    :param m: int, > 0
+        The dimension of x
+    :param n: int, > 0
+        Number of polynomials to compute
+    :param alpha: float, > -1
+    :param beta: float, > -1
+    :param x: (m,) ndarray
+    :return: v: (m, n + 1) ndarray
+        v[:, i] = P^{(alpha, beta)}_n(x) as defined in the paper
+    """
+
+    if n < 0:
+        return np.array([])
+    if n == 0:
+        return np.ones(m)
+    x = x.reshape((m,))
+    v = np.zeros((m, n + 1))
+    alpha_p_beta = alpha + beta
+    alpha_m_beta = alpha - beta
+    v[:, 0] = 1
+    v[:, 1] = (1 + 0.5 * alpha_p_beta) * x + 0.5 * alpha_m_beta
+
+    for i in range(2, n + 1):
+        c1 = 2 * i * (i + alpha_p_beta) * (2 * i + alpha_p_beta - 2)
+        c2 = (2 * i + alpha_p_beta) * (2 * i + alpha_p_beta - 1) * (2 * i + alpha_p_beta - 2)
+        c3 = (2 * i + alpha_p_beta - 1) * alpha_p_beta * alpha_m_beta
+        c4 = -2 * (i - 1 + alpha) * (i - 1 + beta) * (2 * i + alpha_p_beta)
+        v[:, i] = ((c3 + c2 * x) * v[:, i - 1] + c4 * v[:, i - 2]) / c1
+
+    return v
+
+
+def d_decay_approx_fun(a, b, c, d):
+    return np.square(c) / (16 * (np.square(d) + d * (2 * b + a + 1)) - np.square(c))
+
+
+def p_n(n, alpha, beta, x):
+    """
+    wrapper to j_polynomial, returns the first n jacobi polynomial of x
+    """
+    return j_polynomial(len(x), n, alpha, beta, x)
+
+
+def t_x_mat(x, n, j, approx_length):
+    a = np.power(x, n + 0.5)
+    b = np.sqrt(2 * (2 * j + n + 1))
+    c = p_n(approx_length - 1, n, 0, 1 - 2 * np.square(x))
+    return np.einsum('i,j,ij->ij', a, b, c)
+
+
+def t_x_mat2(x, n, j, approx_length):
+    # use np.outer instead of x as mat
```

Comment:
It is different to t_x_mat and only used in the fast PSWF method. The input of x in t_x_mat2 need to be a matrix instead of a vector in t_x_mat. I renamed t_x_mat2 to t_x_mat_dot according to the dot operation in it.